File: LeetCode/Medium/0011-container-with-most-water/0011-container-with-most-water.java
class Solution {
    public int maxArea(int[] height) {

        int left = 0;
        int right = height.length - 1;

        int water = 0;
        int w = 0;
        int h = 0;
        int max = Integer.MIN_VALUE;
        while(left < right) {
            int LHeight = height[left];
            int RHeight = height[right];
            w = right - left;
            h = Math.min(LHeight, RHeight);
            water = w * h;
            max = Math.max(water, max);

            if(LHeight < RHeight) {
                left++;
            } else {
                right--;
            }
        }

        return max;
    }
}

/*
    물의 높이 구하려면
    양 쪽의 기둥을 보고 그 중 작은쪽을 기준으로 height가 돼서 물이 찬다
    width는 두 막대간의 idx차이
    
    w = height[right] - height[left]
    h = Math.min(height[left], height[right])
    water = w * h

    하나하나 다 찾아줄 수 없으니 투포인터로 n스캔만에 찾아줄거임
    
    height[left] < height[right] 면, 물 구해준다음에
    다음걸로 비교 이어나가야하므로 작은쪽인 left를 이동(++)
    반대면 right--
*/

File: LeetCode/Medium/0011-container-with-most-water/0011-container-with-most-water.ts
function maxArea(height: number[]): number {
    let left:number = 0;
    let right:number = height.length - 1;
    let result = -Infinity;

    let water:number = 0;
    while(left < right) {
        const w = right - left;
        const h = Math.min(height[left], height[right]);
        water = w * h;
        result = Math.max(result, water);

        if(height[left] < height[right]) {
            left++;
        } else {
            right--;
        }
    }

    return result;
}; 


/*
    물 높이를 구하는 방법은
    좌/우 비교헀을때 작은쪽 기준이 높이가되고
    둘 사이의 인덱스 차이가 넓이가 됨
    따라서 min(h[right], h[left]) 랑 right-left를 곱하면 그 right,left값의 water가 됨

    투포인터로 구간을 다 구해서 max를 찾아줄건데
    포인터가 양쪽에서 좁혀가면서, 이동조건은 높이가 작은쪽의 높이ㄷ쪽 포인터를 땅겨줄거임
    거긴 이미 계산했으니까 새로운 계산을 하기 위함임
*/


